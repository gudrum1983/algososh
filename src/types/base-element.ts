import {CircleProps} from "../components/ui/circle/circle";
import {ColumnProps} from "../components/ui/column/column";

export type CircleBaseElement = Required<CircleProps> & {id: string}

export type ColumnBaseElement = Required<ColumnProps> & {id: string}


