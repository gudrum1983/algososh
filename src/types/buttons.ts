export enum Buttons {
  addHead = "addHead",
  addTail = "addTail",
  deleteHead = "deleteHead",
  deleteTail = "deleteTail",
  addByIndex = "addByIndex",
  deleteByIndex = "deleteByIndex",
  clear = "clear",
  reverse = "reverse",
  fibonacci = "fibonacci",
  sortDescending = "sortDescending",
  sortAscending = "sortAscending"
}