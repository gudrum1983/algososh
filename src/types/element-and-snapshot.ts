import {CircleProps} from "../components/ui/circle/circle";
import {ColumnProps} from "../components/ui/column/column";

export type CircleBaseElement = Required<CircleProps> & {id: string}

export type ColumnBaseElement = Required<ColumnProps> & {id: string}

export type TSnapshot<T> = Array<T>

export type TSnapshotsList<T> = Array<Array<T>>

