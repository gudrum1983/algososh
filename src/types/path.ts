export enum Path {
  string = "/string",
  fibonacci = "/fibonacci",
  sorting = "/sorting",
  stack = "/stack",
  queue = "/queue",
  list = "/list",
}