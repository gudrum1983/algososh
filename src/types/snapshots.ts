export interface ISnapshot<T> {
  getState(): T;
}

export interface ISnapshotStorage<T> {
  retrieveAndRemoveSnapshot: () => T | null;
  createAndStoreSnapshot: () => void;
  clear: () => void;
  isEmpty:() => boolean;
}
