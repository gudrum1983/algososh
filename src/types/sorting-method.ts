export const BUBBLE = "BUBBLE"
export const SELECTION = "SELECTION"

export type TSortingMethod = typeof BUBBLE | typeof SELECTION