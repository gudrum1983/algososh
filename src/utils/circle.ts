import {ElementStates} from "../types/element-states";
import {nanoid} from "nanoid";

export interface ICircleElement {
  setDefaultState: () => void;
  setChangingState: () => void;
  setModifiedState: () => void;
  setLetter: (value: string | number) => void;
  clearLetter: () => void;
  setHead: (head: string | React.ReactElement) => void;
  clearHead: () => void;
  setTail: (tail: string | React.ReactElement) => void;
  clearTail: () => void;
  getIndex: () => number;
  setIndex: (index: number) => void;
  setSmallSize: (isSmall: boolean) => void;
  getElement:() => IStateCircleElement;
}

export interface IStateCircleElement {
  state: ElementStates;
  letter?: string;
  head?: string | React.ReactElement | null;
  index: number;
  tail?: string | React.ReactElement | null;
  isSmall?: boolean;
  id: string;
}

export class CircleElement implements ICircleElement {
  private state: ElementStates;
  private letter?: string;
  private head?: string | React.ReactElement | null;
  private index: number;
  private tail?: string | React.ReactElement | null;
  private isSmall?: boolean;
  private readonly id: string;

  constructor(value: string | number, index: number, id?: string) {
    this.state = ElementStates.Default;
    this.letter = String(value);
    this.head = null;
    this.index = index;
    this.tail = null;
    this.isSmall = false;
    this.id = id ? id :nanoid(5)
  }

  setDefaultState() {
    this.state = ElementStates.Default
  };

  setChangingState() {
    this.state = ElementStates.Changing
  };

  setModifiedState() {
    this.state = ElementStates.Modified
  };

  setLetter(value: string | number) {
    this.letter = String(value)
  };

  clearLetter() {
    this.letter = ""
  };

  setHead(head: string | React.ReactElement) {
    this.head = head
  };

  clearHead() {
    this.head = null
  };

  setTail(tail: string | React.ReactElement) {
    this.tail = tail
  };

  clearTail() {
    this.tail = null
  };

  getIndex() {
    return this.index
  };

  setIndex(index: number) {
    this.index = index
  };

  setSmallSize(isSmall: boolean) {
    this.isSmall = isSmall
  };

  getElement(): IStateCircleElement {
    return {
      state: this.state,
      letter: this.letter,
      head: this.head,
      index: this.index,
      tail: this.tail,
      isSmall: this.isSmall,
      id: this.id,
    }
  }
}