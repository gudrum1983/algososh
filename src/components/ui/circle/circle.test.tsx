import React from 'react';
import {Circle} from "./circle";
import {ElementStates} from "../../../types/element-states";

const TestRenderer = require('react-test-renderer');

describe('test Circle component', () => {

  const text: string = "text";
  const head: string = "head";
  const tail: string = "tail";
  const index: number = 33;
  const defState: ElementStates = ElementStates.Default;
  const changState: ElementStates = ElementStates.Changing;
  const modState: ElementStates = ElementStates.Modified;


  it('should render a circle without letter', () => {
    const id: string = "CircleWithoutLetter";
    const tree = TestRenderer
      .create(<Circle
        data-testid={id}/>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render a circle with letter', () => {
    const id: string = "CircleWithLetter";
    const tree = TestRenderer
      .create(<Circle
        data-testid={id}
        letter={text}/>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render a circle with a text header', () => {
    const id: string = "CircleWithTextHead";
    const tree = TestRenderer
      .create(<Circle
        data-testid={id}
        letter={text}
        head={head}/>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render a circle with a circle header', () => {
    const id: string = "CircleWithCircleHead";
    const tree = TestRenderer
      .create(<Circle
        data-testid={id}
        letter={text}
        head={<Circle letter={head} isSmall={true} state={ElementStates.Changing}/>}/>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render a circle with a text tail', () => {
    const id: string = "CircleWithTextTail";
    const tree = TestRenderer
      .create(<Circle
        data-testid={id}
        letter={text}
        tail={tail}/>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render a circle with a circle tail', () => {
    const id: string = "CircleWithCircleTail";
    const tree = TestRenderer
      .create(<Circle
        data-testid={id}
        letter={text}
        head={<Circle letter={tail} isSmall={true} state={changState}/>}/>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render a circle with index', () => {
    const id: string = "CircleWithIndex";
    const tree = TestRenderer
      .create(<Circle
        data-testid={id}
        letter={text}
        index={index}/>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render a small circle', () => {
    const id: string = "SmallCircle";
    const tree = TestRenderer
      .create(<Circle
        data-testid={id}
        letter={text}
        isSmall={true}/>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render a circle in default state', () => {
    const id: string = "DefaultCircle";
    const tree = TestRenderer
      .create(<Circle
        data-testid={id}
        letter={text}
        state={defState}
      />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render a circle in changing state', () => {
    const id: string = "ChangingCircle";
    const tree = TestRenderer
      .create(<Circle
        data-testid={id}
        letter={text}
        state={changState}
      />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render a circle in modified state', () => {
    const id: string = "ModifiedCircle";
    const tree = TestRenderer
      .create(<Circle
        data-testid={id}
        letter={text}
        state={modState}
      />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});