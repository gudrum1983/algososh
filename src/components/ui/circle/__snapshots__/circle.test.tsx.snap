// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Circle component should render a circle in changing state 1`] = `
<div
  className="content "
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head element"
  />
  <div
    className="circle   changing"
  >
    <p
      className="text text_type_circle text_color_input letter"
    >
      text
    </p>
  </div>
  <p
    className="text text_type_input text_color_input mt-4 absolute index"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail30 element"
  />
</div>
`;

exports[`test Circle component should render a circle in default state 1`] = `
<div
  className="content "
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head element"
  />
  <div
    className="circle   default"
  >
    <p
      className="text text_type_circle text_color_input letter"
    >
      text
    </p>
  </div>
  <p
    className="text text_type_input text_color_input mt-4 absolute index"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail30 element"
  />
</div>
`;

exports[`test Circle component should render a circle in modified state 1`] = `
<div
  className="content "
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head element"
  />
  <div
    className="circle   modified"
  >
    <p
      className="text text_type_circle text_color_input letter"
    >
      text
    </p>
  </div>
  <p
    className="text text_type_input text_color_input mt-4 absolute index"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail30 element"
  />
</div>
`;

exports[`test Circle component should render a circle with a circle header 1`] = `
<div
  className="content "
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head element"
  >
    <div
      className="content "
    >
      <div
        className="text text_type_input text_color_input mb-4 absolute head element"
      />
      <div
        className="circle  small changing"
      >
        <p
          className="text text_type_circle text_color_input letter"
        >
          head
        </p>
      </div>
      <p
        className="text text_type_input text_color_input mt-4 absolute index"
      />
      <div
        className="text text_type_input text_color_input mt-4 absolute tail30 element"
      />
    </div>
  </div>
  <div
    className="circle   default"
  >
    <p
      className="text text_type_circle text_color_input letter"
    >
      text
    </p>
  </div>
  <p
    className="text text_type_input text_color_input mt-4 absolute index"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail30 element"
  />
</div>
`;

exports[`test Circle component should render a circle with a circle tail 1`] = `
<div
  className="content "
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head element"
  >
    <div
      className="content "
    >
      <div
        className="text text_type_input text_color_input mb-4 absolute head element"
      />
      <div
        className="circle  small changing"
      >
        <p
          className="text text_type_circle text_color_input letter"
        >
          tail
        </p>
      </div>
      <p
        className="text text_type_input text_color_input mt-4 absolute index"
      />
      <div
        className="text text_type_input text_color_input mt-4 absolute tail30 element"
      />
    </div>
  </div>
  <div
    className="circle   default"
  >
    <p
      className="text text_type_circle text_color_input letter"
    >
      text
    </p>
  </div>
  <p
    className="text text_type_input text_color_input mt-4 absolute index"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail30 element"
  />
</div>
`;

exports[`test Circle component should render a circle with a text header 1`] = `
<div
  className="content "
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head string"
  >
    head
  </div>
  <div
    className="circle   default"
  >
    <p
      className="text text_type_circle text_color_input letter"
    >
      text
    </p>
  </div>
  <p
    className="text text_type_input text_color_input mt-4 absolute index"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail30 element"
  />
</div>
`;

exports[`test Circle component should render a circle with a text tail 1`] = `
<div
  className="content "
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head element"
  />
  <div
    className="circle   default"
  >
    <p
      className="text text_type_circle text_color_input letter"
    >
      text
    </p>
  </div>
  <p
    className="text text_type_input text_color_input mt-4 absolute index"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail30 string"
  >
    tail
  </div>
</div>
`;

exports[`test Circle component should render a circle with index 1`] = `
<div
  className="content "
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head element"
  />
  <div
    className="circle   default"
  >
    <p
      className="text text_type_circle text_color_input letter"
    >
      text
    </p>
  </div>
  <p
    className="text text_type_input text_color_input mt-4 absolute index"
  >
    33
  </p>
  <div
    className="text text_type_input text_color_input mt-4 absolute tail60 element"
  />
</div>
`;

exports[`test Circle component should render a circle with letter 1`] = `
<div
  className="content "
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head element"
  />
  <div
    className="circle   default"
  >
    <p
      className="text text_type_circle text_color_input letter"
    >
      text
    </p>
  </div>
  <p
    className="text text_type_input text_color_input mt-4 absolute index"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail30 element"
  />
</div>
`;

exports[`test Circle component should render a circle without letter 1`] = `
<div
  className="content "
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head element"
  />
  <div
    className="circle   default"
  >
    <p
      className="text text_type_circle text_color_input letter"
    />
  </div>
  <p
    className="text text_type_input text_color_input mt-4 absolute index"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail30 element"
  />
</div>
`;

exports[`test Circle component should render a small circle 1`] = `
<div
  className="content "
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head element"
  />
  <div
    className="circle  small default"
  >
    <p
      className="text text_type_circle text_color_input letter"
    >
      text
    </p>
  </div>
  <p
    className="text text_type_input text_color_input mt-4 absolute index"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail30 element"
  />
</div>
`;
